---
title: "mapping-workshop"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# load libraries and data
library(qtl2)
sessionInfo()
iron <- read_cross2(file = system.file("extdata", "iron.zip", package = "qtl2"))

# look at a summary of the iron data
summary(iron)
```

```{r}
names(iron)
head(iron$gmap)
```

```{r}
# insert pseudomarkers at 1 cM intervals
map <- insert_pseudomarkers(map = iron$gmap, step = 1)
head(map, n = 2)
```

```{r}
pr <- calc_genoprob(cross = iron, map = map, error_prob = 0.002)
names(pr)
dimnames(pr$`19`)
```

```{r}
# look at genotype probabilities
(pr$`19`)[1:3,,"D19Mit68"]
```

```{r}
Xcovar <- get_x_covar(iron)
View(Xcovar)
```

Perform a genome scan.

```{r}
out <- scan1(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar)
```


```{r}
head(out)
```

```{r}
head(out, n=30)
```

Chromosomes 2 and 16 for liver appear to contain QTL.
```{r}
sort(out[ , 1])
sort(out[ , "liver"])
```

What about for spleen?

```{r}
sort(out[ , 2], decreasing = TRUE)
sort(out[ , "spleen"])

```

Plot the LOD scores for spleen.
```{r}
plot_scan1(out, map = map, lodcolumn = "spleen")
```

How do we know how a large a LOD score is significant?

```{r}
operm <- scan1perm(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar, n_perm = 10)
summary(operm)
```
Obtain X-chromosome specific thresholds.
```{r}
operm2 <- scan1perm(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar, n_perm = 10, perm_Xsp = TRUE, chr_lengths = chr_lengths(map))
summary(operm2, alpha = c(0.2, 0.05))
```

```{r}
shuffled_order <- sample(rownames(iron$pheno))
pheno_permuted <- iron$pheno
rownames(pheno_permuted) <- shuffled_order
xcovar_permuted <- Xcovar
rownames(xcovar_permuted) <- shuffled_order
out_permuted <- scan1(genoprobs = pr, pheno = pheno_permuted, Xcovar = xcovar_permuted)
plot(out_permuted, map)
head(iron$pheno, n = 3)
head(pheno_permuted, n = 3)
```

Finding LOD peaks

```{r}
summary(operm)
thr <- summary(operm)
find_peaks(scan1_output = out, map = map, threshold = thr, prob = 0.95, expand2markers = FALSE)
```

```{r}
find_peaks(scan1_output = out, map = map, threshold = thr, prob = 0.90, expand2markers = FALSE)
```


```{r}
kinship <- calc_kinship(probs = pr)
kinship[1:5, 1:5]
heatmap(kinship)
```
```{r}
kinship_loco <- calc_kinship(probs = pr, type = "loco")
```

```{r}
out <- scan1(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar)
out_pg <- scan1(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar, kinship = kinship)
out_loco <- scan1(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar, kinship = kinship_loco)
```

```{r}
plot_scan1(out_loco, map = map, lodcolumn = "liver", col = "black")
plot_scan1(out_pg, map = map, lodcolumn = "liver", col = "blue", add = TRUE)
plot_scan1(out, map = map, lodcolumn = "liver", add = TRUE, col = "green")
```

```{r}
file <- paste0("https://raw.githubusercontent.com/rqtl/",
"qtl2data/master/B6BTBR/b6btbr.zip")
b6btbr <- read_cross2(file)
summary(b6btbr)
head(b6btbr$pheno)
colnames(b6btbr$pheno)
b6bmap <- insert_pseudomarkers(map=b6btbr$gmap, step=1)
prb6b <- calc_genoprob(cross=b6btbr, map=b6bmap, error_prob=0.002)
b6bXcovar <- get_x_covar(b6btbr)
b6bout <- scan1(genoprobs = prb6b, pheno = b6btbr$pheno, Xcovar=b6bXcovar)
plot(b6bout, map = b6bmap)
b6boperm <- scan1perm(prb6b, b6btbr$pheno[,1], Xcovar=b6bXcovar, n_perm=10)
summary(b6boperm)
b6bthr = summary(b6boperm)
find_peaks(scan1_output = b6bout, map = b6bmap, threshold = b6bthr, prob = 0.95, expand2markers = FALSE)
b6bkinship <- calc_kinship(probs = prb6b)
out_pg_b6b <- scan1(prb6b, b6btbr$pheno, kinship=b6bkinship, Xcovar=b6bXcovar)
kinship_loco_b6b <- calc_kinship(prb6b, "loco")
out_pg_loco_b6b <- scan1(prb6b, b6btbr$pheno, kinship_loco_b6b, Xcovar=b6bXcovar)
plot_scan1(out_pg_loco_b6b, map = b6bmap, lodcolumn = "log10_insulin_10wk", col = "black")
plot_scan1(out_pg_b6b, map = b6bmap, lodcolumn = "log10_insulin_10wk", col = "blue", add = TRUE)
plot_scan1(b6bout, map = b6bmap, lodcolumn = "log10_insulin_10wk", col = "green", add = TRUE)
```

Estimating QTL effects

```{r}
c2eff <- scan1coef(genoprobs = pr[,"2"], pheno = iron$pheno[, "liver"])
head(c2eff)
```

```{r}
plot(c2eff, map["2"], columns = 1:3, col = c("blue", "red", "green"))
```

```{r}
c2eff_pg <- scan1coef(genoprobs = pr[, "2"], pheno = iron$pheno[, "liver"], kinship_loco[["2"]])
head(c2eff_pg)
```

```{r}
plot(c2eff_pg, map["2"], columns = 1:3, col = c("blue", "red", "green"), ylab = "phenotype average")

```

Best linear unbiased predictor
```{r}
c2blup <- scan1blup(pr[,"2"], iron$pheno[, "liver"], kinship_loco[["2"]])
```

```{r}
plot(c2eff, map["2"], columns = 1:3, col = c("blue", "red", "green"), xlab="Chr 2 position")
plot(c2blup, map["2"], columns = 1:3, col = c("blue", "red", "green"), add=TRUE, lty=2)
```





