---
title: "Performing a permutation test"
teaching: 10
exercises: 20
questions:
- "How can I evaluate the statistical significance of genome scan results?"
objectives:
- Run a permutation test to establish LOD score thresholds.
keypoints:
- "A permutation test establishes the statistical  significance of a genome scan."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

```{r load_dependencies_06, echo=FALSE}
library(qtl2)
iron <- read_cross2( system.file("extdata", "iron.zip", package="qtl2") )
map <- insert_pseudomarkers(map=iron$gmap, step=1)
pr <- calc_genoprob(cross=iron, map=map, error_prob=0.002)
Xcovar <- get_x_covar(iron)
bin_pheno <- apply(iron$pheno, 2, function(a) as.numeric(a > median(a)))
rownames(bin_pheno) <- rownames(iron$pheno)

# run code/run_perms.R to create the next 3 data objects
load("../data/operm.Rdata")
load("../data/operm2.Rdata")
load("../data/operm_bin.Rdata")
```

To establish the statistical significance of the results of a genome scan, a permutation test can identify the maximum LOD score that can occur by random chance. A permutation tests shuffles genotypes and phenotypes, essentially breaking the relationship between the two. The genome-wide maximum LOD score is then calculated on the permuted data, and this score used as a threshold of statistical significance. A genome-wide maximum LOD on shuffled, or permuted, data serves as the threshold because it represents the highest LOD score generated by random chance. 

The `scan1perm()` function takes the same arguments as `scan1()`, plus additional arguments to control the permutations:

- `n_perm` is the number of permutation replicates.
- `perm_Xsp` controls whether to perform autosome/X chromosome specific permutations (with `perm_Xsp=TRUE`) or not (the default is FALSE).
- `perm_strata` is a vector that defines the strata for a stratified permutation test.
- `chr_lengths` is a vector of chromosome lengths, used in the case that `perm_Xsp=TRUE`.

As with `scan1()`, you may provide a kinship matrix (or vector of kinship matrices, for the "leave one chromosome out" (loco) approach), in order to fit linear mixed models. If `kinship` is unspecified, the function performs ordinary Haley-Knott regression.

To perform a permutation test with the `iron` data, we run `scan1perm()`, provide it with the genotype probabilities, the phenotype data, X covariates and number of permutations. For expediency, we'll use only 10 permutations, although 1000 is recommended.

```{r scan1perm, eval = FALSE}
operm <- scan1perm(genoprobs = pr, pheno = iron$pheno, Xcovar = Xcovar, n_perm = 1000) # replace 1000 with 10 for expediency
```

Note the need to specify special covariates for the X chromosome (via `Xcovar`), to be included under the null hypothesis of no QTL. And note that when these are provided, the default is to perform a stratified permutation test, using strata defined by the rows in
`Xcovar`. In general, when the X chromosome is considered, one will wish to stratify at least by sex.

Also note that, as with `scan1()`, you can speed up the calculations on a multi-core machine by specifying the argument `cores`. With `cores=0`, the number of available cores will be detected via `parallel::detectCores()`. Otherwise, specify the number of cores as a positive integer. For large datasets, be mindful of the amount of memory that will be needed; you may need to use fewer than the maximum number of cores, to avoid going beyond the available memory.

```{r scan1perm_multicore, eval=FALSE}
operm <- scan1perm(pr, iron$pheno, Xcovar=Xcovar, n_perm=1000, cores=0)
```

`operm` now contains the maximum LOD score for each permutation for the liver and spleen phenotypes. There should be 1000 values for each phenotypes. We can view the liver permutation LOD scores by making a histogram.

```{r hist_perm}
hist(operm[,'liver'], breaks = 50, xlab = "LOD", main = "LOD scores for liver scan with threshold in red")
abline(v = summary(operm)[,'liver'], col = 'red', lwd = 2)
```

In the histogram above, you can see that most of the maximum LOD scores fall between 1 and 3. This means that we expect LOD scores less than 3 to occur by chance fairly often. The red line indicates the alpha = 0.05 threshold, which means that, under permutation, we only see LOD values as high or higher, 5% of the time. This is one way of estimating a significance threshold for QTL plots.

To get estimated significance thresholds, use the function `summary()`.

```{r summary_scan1perm}
summary(operm)
```

The default is to return the 5% significance thresholds. Thresholds for other (or for multiple) significance levels can be obtained via the `alpha` argument.

```{r summary_scan1perm_B}
summary(operm, alpha=c(0.2, 0.05))
```

To obtain autosome/X chromosome-specific significance thresholds, specify `perm_Xsp=TRUE`. In this case, you need to provide chromosome lengths, which may be obtained with the function `chr_lengths()`.

```{r scan1perm_Xsp, eval=FALSE}
operm2 <- scan1perm(pr, iron$pheno, Xcovar=Xcovar, n_perm=1000,
                    perm_Xsp=TRUE, chr_lengths=chr_lengths(map))
```

Separate permutations are performed for the autosomes and X chromosome, and considerably more permutation replicates are needed for the X chromosome. The computations take about twice as much time.
See [Broman et al. (2006) Genetics
174:2151-2158](https://www.ncbi.nlm.nih.gov/pubmed/17028340).

The significance thresholds are again derived via `summary()`:

```{r summary_scan1perm_C}
summary(operm2, alpha=c(0.2, 0.05))
```

As with `scan1`, we can use `scan1perm` with binary traits, using the argument `model="binary"`. Again, this can't be used with a kinship matrix, but all of the other arguments can be applied.

```{r scan1perm_binary, eval=FALSE}
operm_bin <- scan1perm(pr, bin_pheno, Xcovar=Xcovar, n_perm=1000, 
                       perm_Xsp=TRUE, chr_lengths=chr_lengths(map),
                       model="binary")
```

Here are the estimated 5% and 20% significance thresholds.

```{r summary_scan1perm_binary}
summary(operm_bin, alpha=c(0.2, 0.05))
```

The code below shuffles the phenotypes so that they no longer match up with the genotypes. The purpose of this is to find out how high the LOD score can be due to random chance alone.

```{r permute_data, eval=FALSE}
shuffled_order <- sample(rownames(iron$pheno))
pheno_permuted <- iron$pheno
rownames(pheno_permuted) <- shuffled_order
xcovar_permuted <- Xcovar
rownames(xcovar_permuted) <- shuffled_order
out_permuted <- scan1(genoprobs = pr, pheno = pheno_permuted, Xcovar = xcovar_permuted)
plot(out_permuted, map)
head(shuffled_order)
```

> ## Challenge 1
> Run the preceding code to shuffle the phenotype data and plot a genome
> scan with this shuffled (permuted) data.  
>
> What is the maximum LOD score in the scan from this permuted data?  
> How does it compare to the maximum LOD scores obtained from the earlier scan?  
> How does it compare to the 5% and 20% LOD thresholds obtained earlier?  
> Paste the maximum LOD score in the scan from your permuted data into the etherpad.
>
> > ## Solution to Challenge 1
> >
> {: .solution}
{: .challenge}

> ## Challenge 2
> 1) Find the 1% and 10% significance thresholds for the first set of 
permutations contained in the object `operm`.  
> 2) What do the 1% and 10% significance thresholds say about LOD scores?
> 
> > ## Solution to Challenge 2
> > 1) `summary(operm, alpha=c(0.01, 0.10))`  
> > 2) These LOD thresholds indicate maximum LOD scores that can be obtained
by random chance at the 1% and 10% significance levels. We expect to see LOD 
values this high or higher 1% and 10% of the time respectively.
> {: .solution}
{: .challenge}
