---
title: "QTL analysis in Diversity Outbred Mice"
teaching: 0
exercises: 0
questions:
- "?"
objectives:
- 
- 
keypoints:
- "."
- "."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("12-")
```

DO QTL Mapping Tutorial
========================================================
Introduction
--------------------------------------------------------

This tutorial will take you through the process of mapping a QTL and searching for candidate genes. It assumes that you have already phenotyped and genotyped your mice, and have reconstructed their genomes using the DOQTL calc.genoprob() function. DOQTL runs in the R statistical package and this demo assumes some familiarity with that software. You should have been provided the following files for the demo:

* DOQTL_demo.Rdata: contianing the phenotype and genotype data.

The data comes from a toxicology study in which mice were exposed to benzene via inhalation for 6 hours a day, 5 days a week for 4 weeks. The study was conducted in two equally sized cohort of 300 male mice each, for a total of 600 mice. They were then sacrificed and reticulocytes (red blood cell precursors) were isolated from bone marrow. The number of micro-nucleated reticulocytes, a measure of DNA damage, was then measured in each mouse. The goal is to map gene(s) that influence the level of DNA damage in the bone marrow.

![](figure/benzene_study_design.png)

Loading the DOQTL package
--------------------------------------------------------

The following code loads DOQTL's dependencie and installs DOQTL.  You do not need to run it now, but this is what you would run to install DOQTL on your computer at home.  Note that the installation of dependencies takes a while becuase the mouse and human genomes are downloaded.

   source("http://bioconductor.org/biocLite.R")  
   biocLite("DOQTL")

First, we load DOQTL into the R environment.

```{r library, message=FALSE, results='hide', eval=FALSE}
library(DOQTL)
```

Loading the data
--------------------------------------------------------

The data for this tutorial has been saved as an R binary file that contains several data objects.  Load it in now by running the following command.

```{r load_data,message=FALSE,results='hide'}
load("/data/DOQTL_demo.Rdata")
```

This loaded in two data objects. Look in the Global Environment panel to see what was loaded.  You should see an obejct called `pheno` with 143 rows and 5 columns and an object called `probs`.

`pheno` is a data frame containing the phenotype data. `probs` is a 3 dimensional array containing the founder allele dosages for each sample at each marker on the array.  Double-click on `pheno` in the Global Environment panel to view its contents.

**NOTE:** the sample IDs must be in the rownames of `pheno`.

It contains the sample ID, the study cohort, the dose of benzene and the proportion of bone marrow reticulocytes that were micro-nucleated (prop.bm.MN.RET).  Note that the sample IDs are also stored in the rownames of pheno. In order to save time for this tutorial, we will only map with 143 samples from the 100 ppm dosing group.

Next, we look at the contents of `probs`:

```{r}
dim(probs)
```

`probs` is a three dimensional array containing the proportion of each founder haplotype at each marker for each DO sample.  The 143 samples are in the first dimension, the 8 founders in the second and the markers along the mouse genome are in the third dimension. Let's look at the contents for the first 500 markers of one sample.

**NOTE:** the sample IDs must be in the rownames of `probs`.

```{r geno_plot,fig.width=8, fig.height=6}
image(1:500, 1:ncol(probs), t(probs[1,8:1,1:500]), breaks = 0:100/100,
      col = grey(99:0/100), axes = F, xlab = "Markers", ylab = "Founders",
      main = "Founder Allele Contributions for Sample 1")
abline(h = 0:8 + 0.5, col = "grey70")
usr = par("usr")
rect(usr[1], usr[3], usr[2], usr[4])
axis(side = 1, at = 0:5 * 100, labels = 0:5 * 100)
axis(side = 2, at = 1:8, labels = LETTERS[8:1], las = 1, tick = F)
```

In the plot above, the founder contributions, which range between 0 and 1, are colored from white (= 0) to black (= 1.0). A value of ~0.5 is grey. The markers are on the X-axis and the eight founders (denoted by the letters A through H) on the Y-axis. Starting at the left, we see that this sample has genotype CD because both rows C and D are grey, indicating values o 0.5 for each one. Moving along the genome to the right, the genotype becomes DD where row D is black, then CD, AC, CH, CD, CH, etc. The value at each marker sum to 1.0.

QTL Mapping
-------------------------------------------------------

First, we need the locations of the markers on the genotyping array. The array is called the Mouse Universal Genotyping Array (MUGA) and contain 7,856 SNP markers. Their locations are on [The Jackson Laboratory's FTP site](ftp://ftp.jax.org/MUGA):

```{r snps}
load(url("ftp://ftp.jax.org/MUGA/muga_snps.Rdata"))
```

Next, we need to create a matrix that accounts for the kinship relationships between the mice. We do this by looking at the correlation between the founder haplotypes for each sample at each SNP. For each chromosome, we create a kinship matrix using the all markers *except* the ones on the current chromosome. Simulations suggest that mapping using this approach increases the power to detect QTL.
           
```{r kinship,message=FALSE,results='hide'}
K = kinship.probs(probs, snps = muga_snps, bychr = TRUE)
```

Kinship values between pairs of samples range between 0 (no relationship) and 1.0 (completely identical). Let's look at the kinship matrix.

```{r kinship_probs,fig.width=8,fig.height=8}
image(1:nrow(K[[1]]), 1:ncol(K[[1]]), K[[1]][,ncol(K[[1]]):1], xlab = "Samples", 
      ylab = "Samples", yaxt = "n", main = "Kinship between samples", 
      breaks = 0:100/100, col = heat.colors(length(0:100) - 1))
axis(side = 2, at = 20 * 0:7, labels = 20 * 7:0, las = 1)
```

The figure above shows kinship between all pairs of samples. White (= 1) indicates no kinship and red (= 0) indicates full kinship. Orange values indicate varying levels of kinship between 0 and 1. The white diagonal of the matrix indicates that each sample is identical to itself. The lighter yellow blocks off of the diagonal may indicate siblings or cousins.

Next, we need to create additive covariates that wil be used in the mapping model. We will use sex and study cohort as a covariate in the mapping model.  While all of the samples We must add the sample IDs to the rownames of the covariates becuase the 'scanone' function will match up sample IDs in all of the data.

```{r covariates}
addcovar = model.matrix(~Study, data = pheno)
colnames(addcovar)[1] = "Sex"
```         

The code above copies the rownames(pheno) to rownames(addcovar).

**NOTE:** the sample IDs must be in the rownames of `addcovar`.

In order to map prop.bm.MN.RET, you will use the scanone() function. To see the arguments for `scanone`, you can type 'help(scanone)'.

```{r QTL,warning=FALSE,error=FALSE, eval=FALSE}
qtl = scanone(pheno = pheno, pheno.col = "prop.bm.MN.RET", probs = probs, K = K, 
      addcovar = addcovar, snps = muga_snps)
```

We can then plot the QTL scan.         

```{r qtl_plot,fig.width=8, fig.height=6, warning=FALSE,eval=FALSE}
plot(qtl, main = "prop.bm.MN.RET")
```

There is clearly a large peak on Chr 10. Next, we must assess its statistical significance. This is most commonly done via [permutation](http://www.genetics.org/content/178/1/609.long). We advise running at least 1,000 permutations to obtain significance thresholds. In the interest of time, we preform 100 permutations here.

```{r perms,message=FALSE,results='hide', warning=FALSE, eval=FALSE}
perms = scanone.perm(pheno = pheno, pheno.col = "prop.bm.MN.RET", probs = probs,
        addcovar = addcovar, snps = muga_snps, nperm = 100)
```

We can now add thresholds to the previous QTL plot. We use significance thresholds at the p < 0.05, 0.10 and 0.63 levels.
           
```{r qtl_plot_thr,fig.width=8, fig.height=6, warning=FALSE, eval=FALSE}
thr = get.sig.thr(perms[,,1], alpha = c(0.05, 0.1, 0.63), Xchr = FALSE)
plot(qtl, sig.thr = thr, sig.col = c("red", "orange", "goldenrod"), main = "prop.bm.MN.RET")
```

The peak on Chr 10 is clearly well above the red p < 0.05 significance line.

We will now zoom in on Chr 10 and look at the contribution of each of the eight founder alleles to prop.bm.MN.RET. The mapping model fits a term for each of the eight DO founders. We can plot these coefficients across Chr 10.

```{r coef_plot,fig.width=8, fig.height=6, eval=FALSE}
coefplot(qtl, chr = 10, main = "prop.bm.MN.RET")
```

The top panel shows the eight founder allele effects (or model coefficients) along Chr 10. You can see that DO mice containing the CAST/EiJ allele around 34 Mb have lower levels of micro-nucleated reticulocytes. This means that the CAST allele is associated with less DNA damage and has a protective allele. The bottom panel shows the LOD score, with the support interval for the peak shaded blue. The support interval is determined using the [Bayesian Credible Interval](http://www.ncbi.nlm.nih.gov/pubmed/11560912) and represents the region most likely to contain the causative polymorphism(s). We can obtain this interval using the 'bayesint' function.

```{r interval,eval=FALSE}
interval = bayesint(qtl, chr = 10)
knitr::kable(interval)
```

From the table above (column 3), you can see that the support interval is 6 Mb wide (29.7 to 35.7 Mb). Line 1 shows the proximal end of the peak and line 3 shows the distal end. The maximum LOD score and its location are in line 2 of the table.

Next, let's look at the distribution of prop.bm.MN.RET across the 36 possible DO genotypes at the maximum peak location. This can be obtained from the second row of the first column of the interval table (interval[2,1]).

```{r effect_plot,fig.width=8, fig.height=6, eval=FALSE}
pxg.plot(pheno = pheno, pheno.col = "prop.bm.MN.RET", probs = probs, 
         snp.id = interval[2,1], snps = muga_snps)
```

This plot shows the phenotype value on the Y-axis plotted against the 36 DO genotypes on the X-axis, Note that several genotypes did not occur in this set of 143 samples. Next note that all of the samples with at least one CAST allele (denoted by **F**) have low values. This suggests that the mode of inheritance may be dominant. This is the effect that is driving the QTL.

Genome-wide Association Mapping
-------------------------------------------------------------

Above, we performed an analysis called linkage mapping in which we regressed the phenotype on haplotype probabilities.  In this next analysis, we will impute the founder SNPs onto the DO haplotype plots and perform association mapping with 40.4 million SNPs.  For each haplotype block, we copy-and-paste the SNPs from the appropriate founder onto each DO genome.

![](figure/DO.impute.founders.sm.png)

The GWAS function is called `scanone.assoc` and it takes arguments that are similar to `scanone` above. There are two new arguments.  `sdp.file`, which is available from [ftp.jax.org](ftp://ftp.jax.org/MUGA/), is a file that contains compressed SNP information for the 8 DO founder strains.  `ncl` tells the function how many parallel jobs to run to perform the genome scan.  Obviously, don't set this number higher than the number of cores on your computer.  You can find out more about this function using `help(scanone.assoc)`.  

```{r GWAS,warning=FALSE,eval=FALSE}
gwas = scanone.assoc(pheno = pheno, pheno.col = "prop.bm.MN.RET", probs = probs, K = K, 
       addcovar = addcovar, markers = muga_snps,  sdp.file = "/data/DO_Sanger_SDPs.txt.bgz", ncl = 3)
```

Next, we plot the association mapping results.

```{r plot_GWAS,warning=FALSE,eval=FALSE}
plot(gwas, main = "prop.bm.MN.RET")
```

Once again, we should perform permutations of the data to obtain a significance threshold.  This is too time comsuming for this tutorial.  So we will load in pre-computed permutation values and use them to plot significance thresholds.

```{r plot_GWAS_with_thr,warning=FALSE,eval=FALSE}
assoc.perms = readRDS("/data/assoc_perms.rds")
assoc.thr = get.sig.thr(-log10(assoc.perms), Xchr = FALSE)
plot(gwas, sig.thr = assoc.thr, main = "prop.bm.MN.RET")
```

You can also zoom in on one chromosome.

```{r plot_GWAS_chr10,warning=FALSE,eval=FALSE}
plot(gwas, chr = 10, main = "prop.bm.MN.RET")
```

And you can increase the plotting resolution with the `bin.size` argument.

```{r plot_GWAS_chr10_moe_bins,warning=FALSE,eval=FALSE}
plot(gwas, chr = 10, bin.size = 10, main = "prop.bm.MN.RET")
```

Searching for Candidate Genes
-------------------------------------------------------------

At this point, we have a 6 Mb wide support interval that contains a polymorphism(s) that influences benzene induced DNA damage. Next, we will impute the DO founder sequences onto the DO genomes. The [Sanger Mouse Genomes Project](http://www.sanger.ac.uk/resources/mouse/genomes/) has sequenced the eight DO founders and provides SNP, Indel and structural variant files for the strains (see [Baud et.al., Nat. Gen., 2013](http://www.nature.com/ng/journal/v45/n7/full/ng.2644.html)). We can impute these SNPs onto the DO genomes and then perform association mapping. The function `assoc.map` performs this analysis.

```{r assoc_map,warning=FALSE,eval=FALSE}
chr = 10
assoc = assoc.map(pheno = pheno, pheno.col ="prop.bm.MN.RET", probs = probs, K = K[[chr]],
                  addcovar = addcovar, snps = muga_snps, chr = chr, start = interval[1,3],
                  end = interval[3,3], output = "p-value")
```

We can plot the results of the association mapping using assoc.plot().  We set a threshold of 10 to highlight the SNPs with high LOD scores.

```{r assoc_fig,fig.width=9, fig.height=6, results='hide',eval=FALSE}
tmp = assoc.plot(assoc, thr = 10, show.sdps = TRUE)
```

The top panel shows the association of each SNP with MN-RET. Chr 10 is on the X-axis and the LOD score is on the Y-axis. The SNPs for which CAST has an allele that is different from the other 7 founder strains are plotted in red. The bottom panel shows the genes in the interval from Mouse Genome Informatics. There are 60 genes (or non-coding RNAs) in the interval. 

One strategy for finding genes related to a phenotype is to search for genes with expression QTL (eQTL) in the same location. Ideally, we would have liver and bone marrow gene expression data in the DO mice from this experiment. Unfortunately, we did not collect this data. However, we have liver gene expression for a separate set of untreated DO mice [Liver eQTL Viewer](http://cgd.jax.org/apps/eqtlviewer-beta/#). We searched for genes in the QTL interval that had an eQTL in the same location. Then, we looked at the pattern of founder effects to see if CAST stood out. We found two genes that met this criteria.

![](figure/French.et.al.Figure3.png)

As you can see, both *Sult3a1* and *Gm4794* have eQTL in the same location on Chr 10 and mice with CAST allele (in green) express these genes more highly. *Sult3a1* is a sulfotransferase that may be involved in adding a sulphate group to phenol, one of the metabolites of benzene. Go to the Ensembl web page for [Gm4794](http://useast.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=ENSMUSG00000090298;r=10:33766424-33782115;t=ENSMUST00000165904).  In the menu on the left, click on the "Gene Tree (image)" link.

![](figure/EnsEMBL_Sult3a1_Gm4794_paralog.png)

As you can see, *Gm4794* is a paralog of *Sult3a1*. Further research revealed that *Gm4794* contains a sulfotransferase domain.

We also looked at an existing gene expression database in which liver, spleen and kidney gene expression were measured in 26 inbred strains, including the eight DO founders. You can search for *Sult3a1* and *Gm4794* in this [strain survey data](http://cgd.jax.org/gem/strainsurvey26/v1). We did this and plotted the spleen and liver expression values. We did not have bone marrow expression data from this experiment. We also plotted the expression of all of the genes in the QTL support interval that were measured on the array (data not shown).  *Sult3a1* and its paralog *Gm4794* were the only genes with a different expression pattern in CAST. Neither gene was expressed in the spleen.

![](figure/French.et.al.Sup.Figure2.png)

Next, go to the [Sanger Mouse Genomes](http://www.sanger.ac.uk/sanger/Mouse_SnpViewer/rel-1303) website and enter *Sult3a1* into the Gene box. Scroll down and check only the DO founders (129S1/SvImJ, A/J, CAST/EiJ, NOD/ShiLtJ, NZO/HlLtJ & WSB/EiJ) and then scroll up and press 'Search'. This will show you SNPs in *Sult3a1*. Select the 'Structural Variants' tab and note the copy number gain in CAST from 33,764,194 to 33,876,194 bp. Click on the G to see the location, copy this position (10:33764194-33876194) and go to the [Ensembl website](http://useast.ensembl.org/Mus_musculus/Info/Index). Enter the position into the search box and press 'Go'. You will see a figure similar to the one below.

![](figure/EnsEMBL.Sult3a1.png)

Note that both *Gm4794* and part of *Sult3a1* are in the copy number gain region.

In order to visualize the size of the copy number gain, we queried the [Sanger Mouse Genomes alignment files](ftp://ftp-mouse.sanger.ac.uk/current_bams/) for the eight founders. We piled up the reads at each base (which is beyond the scope of this tutorial) and made the figure below.

![](figure/French.et.al.Sup.Figure3.png)

As you can see, there appears to be a duplicatation in the CAST founders that covers four genes: *Clvs2*, *Gm15939*, *Gm4794* and *Sult3a1*. *Clvs2* is expressed in neurons and *Gm15939* is a predicted gene that may not produce a transcript.

Hence, we have three pieces of evidence that narrows our candidate gene list to *Sult3a1* and *Gm4794*:

1. Both genes have a liver eQTL in the same location as the MN-RET QTL.
2. Among genes in the MN-RET QTL interval, only *Sult3a1* and *Gm4794* have differential expression of the CAST allele in teh liver.
3. There is a copy number gain of these two genes in CAST.

This analysis has led us to the following hypothesis. Inhaled benzene is absorbed by the lungs into the blood stream and transported to the liver. There, it is metabolized and some metabolites are transported to the bone marrow. One class of genes that is involved in toxicant metabolism are sulfotransferases. [*Sult3a1*](http://www.informatics.jax.org/marker/MGI:1931469) is a phase II enzyme that conjugates compounds (such as phenol, which is a metabolite of benzene) with a sulfate group before transport into the bile. It is possible that a high level of *Sult3a1* expression could remove benzene by products and be protective. Our hypothesis is that the copy number gain in the CAST allele increases liver gene expression of *Sult3a1* and *Gm4794*. High liver expression of these genes allows mice containing the CAST allele to rapidly conjugate harmful benzene metabolites and excrete them from the body before they can reach the bone marrow and cause DNA damage. Further experimental validation is required, but this is a plausible hypothesis.

![](figure/benzene_hypothesis.png)

We hope that this tutorial has shown you how the DO can be used to map QTL and use the founder effects and bioinformatics resources to narrow down the candidate gene list. Here, we made used of external gene expression data bases and the founder sequence data to build a case for a pair of genes.
